pkg/util/strategicpatch/patch.go

```diff
@@ -827,3 +838,104 @@ func sliceElementType(slices ...[]interface{}) (reflect.Type, error) {
+		patchMap, ok := patch.(map[string]interface{})
+		if !ok {
+			return true
+		}
+
+		// The presence of key means that its value has been changed, so the test fails.
+		_, ok = patchMap[key]
+		return !ok
+	}
+}
+
+// CreateThreeWayMergePatch reconciles a modified configuration with an original configuration,
+// while preserving any changes or deletions made to the original configuration in the interim,
+// and not overridden by the current configuration. All three documents must be passed to the
+// method as json encoded content. It will return a strategic merge patch, or an error if any
+// of the documents is invalid, or if there are any preconditions that fail against the modified
+// configuration, or, if force is false and there are conflicts between the modified and current
+// configurations.
+func CreateThreeWayMergePatch(original, modified, current []byte, dataStruct interface{}, force bool, fns ...PreconditionFunc) ([]byte, error) {
+	originalMap := map[string]interface{}{}
+	err := json.Unmarshal(original, &originalMap)
+	if err != nil {
+		return nil, errBadJSONDoc
+	}
+
+	modifiedMap := map[string]interface{}{}
+	err = json.Unmarshal(modified, &modifiedMap)
+	if err != nil {
+		return nil, errBadJSONDoc
+	}
+
+	currentMap := map[string]interface{}{}
+	err = json.Unmarshal(current, &currentMap)
+	if err != nil {
+		return nil, errBadJSONDoc
+	}
+
+	t, err := getTagStructType(dataStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	// The patch is the difference from current to modified without deletions, plus deletions
+	// from original to modified. To find it, we compute deletions, which are the deletions from
+	// original to modified, and delta, which is the difference from current to modified without
+	// deletions, and then apply delta to deletions as a patch, which should be strictly additive.
+	delta, err := diffMaps(currentMap, modifiedMap, t, false, true)
+	if err != nil {
+		return nil, err
+	}
+
+	deletions, err := diffMaps(originalMap, modifiedMap, t, true, false)
+	if err != nil {
+		return nil, err
+	}
+
+	patch, err := mergeMap(deletions, delta, t)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the preconditions to the patch, and return an error if any of them fail.
+	for _, fn := range fns {
+		if !fn(patch) {
+			return nil, ErrPreconditionFailed
+		}
+	}
+
+	// TODO(jackgr): If the force parameter is false, and the patch contains any keys that are not in
+	// original, then return a conflict error.
+
+	return json.Marshal(patch)
+}
+
+// IsConflicting returns true if the provided error indicates a
+// conflict exists between the original changes and the applied
+// changes.
+func IsConflicting(err error) bool {
+	return err == ErrConflict
+}
+
+// IsPreconditionFailed returns true if the provided error indicates
+// a Delta precondition did not succeed.
+func IsPreconditionFailed(err error) bool {
+	return err == ErrPreconditionFailed
+}
+
+var ErrPreconditionFailed = fmt.Errorf("a precondition failed")
```

Comment:
The ones that are exported are the only ones that should be exported, and they are grouped together.